linbox/solutions/solve/solve-cra.h

```diff
@@ -0,0 +1,190 @@
+        {
+            using FMatrix = typename Matrix::template rebind<Field>::other;
+            using FVector = typename LinBox::Rebind<Vector, Field>::other;
+
+            FMatrix FA(A, F);
+            FVector Fb(F, b);
+
+            LinBox::VectorWrapper::ensureDim(x, A.coldim());
+            return solve(x, FA, Fb, m);
+        }
+    };
+}
+
+namespace LinBox {
+    /**
+     * \brief Solve specialization with Chinese Remainder Algorithm method for an Integer ring.
+     *
+     * If a Dispatch::Distributed is used, please note that the result will only be set on the master node.
+     */
+    template <class Matrix, class Vector, class IterationMethod>
+    inline void solve(Vector& xNum, typename Vector::Element& xDen, const Matrix& A, const Vector& b,
+                      const RingCategories::IntegerTag& tag, const Method::CRA<IterationMethod>& m)
+    {
+        //
+        // Handle auto-dispatch.
+        //
+
+        Dispatch dispatch = m.dispatch;
+        if (dispatch == Dispatch::Auto) {
+            Method::CRA<IterationMethod> newM(m);
+
+#if __LINBOX_HAVE_MPI
+            // User has MPI enabled in config, but not specified if it wanted to use it,
+            // we enable it with default communicator if needed.
+            newM.dispatch = Dispatch::Distributed;
+#else
+            // @note Currently, Sequential = SMP if OpenMP is active on
+            // the machine, as CRA -> ChineseRemainderSequential or ChineseRemainderOMP in cra-domain.h
+            newM.dispatch = Dispatch::Sequential;
+#endif
+
+            return solve(xNum, xDen, A, b, tag, newM);
+        }
+
+        //
+        // Declare communicator if none was yet.
+        //
+
+        if (m.dispatch == Dispatch::Distributed && m.pCommunicator == nullptr) {
+            Method::CRA<IterationMethod> newM(m);
+            Communicator communicator(nullptr, 0);
+            newM.pCommunicator = &communicator;
+            return solve(xNum, xDen, A, b, tag, newM);
+        }
+
+        //
+        // Init all (Hadamard bound, prime generator).
+        //
+
+        if (m.master()) {
+            commentator().start("solve.cra.integer");
+            linbox_check((A.coldim() == xNum.size()) && (A.rowdim() == b.size()));
+        }
+
+        using CraField = Givaro::ModularBalanced<double>;
+        const typename Matrix::Field& F = A.field();
+        unsigned int bits = FieldTraits<CraField>::bestBitSize(A.coldim());
+        PrimeIterator<LinBox::IteratorCategories::HeuristicTag> primeGenerator(bits);
+        CraRebinderSolver<Matrix, Vector, IterationMethod> iteration(A, b, m.iterationMethod);
+
+        // @note The result is stored to Integers, and will be converted
+        // later back.
+        BlasVector<Givaro::ZRing<Integer>> num(F, A.coldim());
+        Integer den(1);
+
+        auto hb = RationalSolveHadamardBound(A, b);
+        double hadamardLogBound = 1.0 + hb.numLogBound + hb.denLogBound; // = ln2(2 * N * D)
```

Comment:
this should be returned by a  dedicated method `solutionLogBound` of `RationalSolveHadamardBound`